src/Features/Core/Portable/DocumentSymbols/AbstractDocumentSymbolsService.cs

```diff
@@ -0,0 +1,185 @@
+
+        private ImmutableArray<DocumentSymbolInfo> GetMembersInDocumentTypesAndMethodsOnly(
+            DocumentSymbolsOptions options,
+            SemanticModel semanticModel,
+            SyntaxTree tree,
+            ISymbolDeclarationService declarationService,
+            CancellationToken cancellationToken)
+        {
+            using var _1 = PooledHashSet<INamedTypeSymbol>.GetInstance(out var typesInFile);
+            GetTypesInFile(semanticModel, typesInFile, cancellationToken);
+
+            using var _2 = ArrayBuilder<DocumentSymbolInfo>.GetInstance(out var typesBuilder);
+            foreach (var type in typesInFile)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                typesBuilder.Add(GetMemberInfoForType(type, tree, declarationService, cancellationToken));
+            }
+
+            typesBuilder.Sort((d1, d2) => d1.Text.CompareTo(d2.Text));
+            return typesBuilder.ToImmutable();
+
+            void GetTypesInFile(SemanticModel semanticModel, HashSet<INamedTypeSymbol> types, CancellationToken cancellationToken)
+            {
+                using (Logger.LogBlock(FunctionId.DocumentSymbols_GetDocumentSymbols_GetTypesInFile, cancellationToken))
+                {
+                    using var _ = ArrayBuilder<SyntaxNode>.GetInstance(out var nodesToVisit);
+
+                    nodesToVisit.Push(semanticModel.SyntaxTree.GetRoot(cancellationToken));
+
+                    while (!nodesToVisit.IsEmpty())
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        var node = nodesToVisit.Pop();
+                        var symbol = GetSymbol(semanticModel, node, cancellationToken);
+
+                        if (symbol is INamedTypeSymbol type)
+                        {
+                            types.Add(type);
+                        }
+
+                        if (ShouldSkipSyntaxChildren(node, options))
+                        {
+                            // quick bail out to prevent us from creating every nodes exist in current file
+                            continue;
+                        }
+
+                        foreach (var child in node.ChildNodes())
+                        {
+                            nodesToVisit.Push(child);
+                        }
+                    }
+                }
+            }
+        }
+
+        private async Task<ImmutableArray<DocumentSymbolInfo>> GetMembersInDocumentFullHierarchyAsync(
+            DocumentSymbolsOptions options,
+            SemanticModel semanticModel,
+            SyntaxTree tree,
+            ISymbolDeclarationService declarationService,
+            CancellationToken cancellationToken)
+        {
+            var root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+
+            using var _ = ArrayBuilder<SyntaxNode>.GetInstance(out var nodesToVisit);
+            nodesToVisit.Push(root);
```

Comment:
Can we add a comment here about what is going on with this shared stack and startingCount?

Reply:
i dont' get it either.